### PR TITLE
Fix for repeated `require()` when `MODULARIZE=1` is used

### LINKED ISSUE #5568: MODULARIZE=1 causes repeated call to fail
I've found an interesting behavior when using `MODULARIZE=1`.
Let's assume `lib.js` is our Emscripten build with `MODULARIZE=1`, then:
```javascript
instance_1 = require('./lib')(); // Everything is good so far
instance_2 = require('./lib')(); //TypeError: require(...) is not a function
```

The things is that during initial call `module.exports` is overridden with plain `Module` without function wrapper.

I've stumbled upon this when writing tests for one project: initially library was called by wrapper code and then I needed to call few internal functions for testing purposes and called `require()` on module directly.

### PR DESCRIPTION
**Pull request by nazar-pc**
Fixes #5568

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -2331,7 +2331,7 @@ def test_llvm_nativizer(self):
     output = Popen([os.path.join(self.get_dir(), 'files.o.run')], stdin=open(os.path.join(self.get_dir(), 'stdin')), stdout=PIPE, stderr=PIPE).communicate()
     self.assertContained('''size: 37
 data: 119,97,107,97,32,119,97,107,97,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35,35
-loop: 119 97 107 97 32 119 97 107 97 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35
```
Comment (kripken) [inline tests/test_other.py]: hmm, that whitespace actually matters. i pushed a fix to incoming now so it doesn't look unnecessary to editors.

File: tests/test_other.py [hunk 2]
```diff
@@ -5438,6 +5438,9 @@ def test_require_modularize(self):
     assert "module['exports'] = NotModule;" in src
     output = Popen(NODE_JS + ['-e', 'var m = require("./a.out.js"); m();'], stdout=PIPE, stderr=PIPE).communicate()
     assert output == ('hello, world!\n', ''), 'expected output, got\n===\nSTDOUT\n%s\n===\nSTDERR\n%s\n===\n' % output
+    Popen([PYTHON, EMCC, path_from_root('tests', 'hello_world.c'), '-s', 'MODULARIZE=1',  '-s', 'NO_EXIT_RUNTIME=1']).communicate()
```
Comment (kripken) [inline tests/test_other.py]: please add a comment saying this tests two calls to `require`

File: src/shell.js [hunk 3]
```diff
@@ -111,9 +111,11 @@ if (ENVIRONMENT_IS_NODE) {
 
   Module['arguments'] = process['argv'].slice(2);
 
+#if MODULARIZE == 0
```
Comment (kripken) [inline src/shell.js]: please add a comment saying why this is not needed in that mode

File: tests/test_other.py (hunk 1 above)
Comment (nazar-pc) [reply tests/test_other.py]: Rebased against incoming

File: tests/test_other.py (hunk 2 above)
Comment (nazar-pc) [reply tests/test_other.py]: Comment added

File: src/shell.js (hunk 3 above)
Comment (nazar-pc) [reply src/shell.js]: Comment added

**Comment by CMCDragonkai** [comment]
I met this issue as well, and my solution to it was not to use `MODULARIZE=1` and instead use prejs and postjs function wrapper that defined a locally scoped `module` to be undefined.

```
var module = undefined;
// or let module = undefined;
```

File: src/shell.js [hunk 4]
```diff
@@ -111,9 +111,12 @@ if (ENVIRONMENT_IS_NODE) {
 
   Module['arguments'] = process['argv'].slice(2);
 
+  // With MODULARIZE == 1 wrapper function will be exported instead of Module itself and re-exporting it second time will cause issues
```
Comment (kripken) [inline src/shell.js]: i still don't fully understand this even with the comment, sorry.

why is the second time special? and why does it matter we export the wrapper function and not the Module? I don't see a connection here that is important, I guess...

Comment (nazar-pc) [reply src/shell.js]: Without this patch `var m = require('module_name')` initially exports wrapper function `module.exports = wrapper_function`, and we need to call this wrapper function to obtain `Module` itself.
However, when `m()` is called it re-exports `module.exports = Module`.
Thus successive `require('module_name')` doesn't return wrapper function anymore, but instead returns plain `Module`.

Just try to run added test on current `incoming` branch.

Comment (curiousdannii) [reply src/shell.js]: There are two `Module` variables. When you `require` it the first time you get the outer wrapper `Module`, and the second time you get the inner true `Module`.

Comment (kripken) [reply src/shell.js]: Oh, I see, so there are two separate places we export, one inside the module - in shell.js, which is disabled by this PR - and another outside the module, added by MODULARIZE, and is the proper place?

If that's correct then looks good to me. Let's add that to the comment, though, something like `MODULARIZE will export the module in the proper place outside, we don't need to export here`.

Comment (nazar-pc) [reply src/shell.js]: That is correct, updated comment accordingly

**Comment by kripken** [comment]
Great, thanks.

**Pull request merged**